### DISCUSSION TITLE
gpsd support?

### DISCUSSION THREAD

**Discussion by WillMarchant** [mcguirepr89/BirdNET-Pi#420] [Ideas]
Has any consideration been given to adding gpsd support so that a mobile installation can get an automatic position and date/time stamp?

We've got the Adafruit "Ultimate GPS Hat" https://www.adafruit.com/product/2324 on our RPis and so it would be great to use those to automatically set up the BirdNET-Pi position and date/time on roving units.
Best wishes,
Will

**Comment by mcguirepr89** [comment]
Hi, @WillMarchant This is a neat and clever idea. I think I would need to actually get my hands on one of those HATs to be able to effectively develop around it. I will put it on my wish-list and hopefully I'll see enough sponsorship support to get that hardware sooner than later.

If you want to start development around that, I would be more than happy getting scripts and tools into the system if you were able to put together some example scripts to parse that info from the HAT?

Perhaps there are other users with similar hardware????
My very best regards,
Patrick

**Comment by WillMarchant** [comment]
Sounds great, Patrick!

Please send me a link to your wishlist, or snaill mail address, and I'll see you get a board.

If wishlist, please add all the "also include" items from the https://www.adafruit.com/product/2324 page.  The GPS antenna mounted on the board works great outdoors but not so well indoors.  The optional external antenna will really make your life easier during development.

That board also requires that the stacking header be soldered onto the board.  A pretty easy task.  But if you aren't comfortable with soldering then let's go the snail-mail route and I'll send you a soldered board.

Please note the caveat on the Adafruit page that the GPS takes over the serial console.  I use http://adafru.it/2260 and http://adafru.it/922 when in the field so that's not a concern for me, but may impact others.

Later today I will post to this discussion a script to start the GPSD daemon running.  And I'll post some example Python code that dumps the lat/lon to standard output.

I have that setup working on v13 but I upgraded to BirdNET-Pi v14 yesterday and want to verify that stuff still works.

**Comment by WillMarchant** [comment]
OK, here are notes on getting the GPS hat interfaced and some Python code to read time, latitude, and longitude.  This does not set the system time.  It could, if NTP got configured, and this https://gpsd.gitlab.io/gpsd/gpsd-time-service-howto.html is probably a place to start.

The setup steps from https://learn.adafruit.com/adafruit-ultimate-gps-hat-for-raspberry-pi/ will get the basic system going.  But here's a summary:
1. Need to get RPi on the network first.  Using wired Ethernet seems easiest.
2. sudo raspi-config to disable serial login but leave serial port hardware enabled
3. sudo apt-get install gpsd gpsd-clients python3-gps
4. load the gpsdsetup.sh script onto the RPi (see below)
5. load the justgps.py program onto the RPi (see below)
6. pip3 install adafruit.blinka
7. pip3 install gpsdclient
8. run ./gpsdsetup.sh to initialize the GPSD server.  We should put this into the boot stream.
9. python justgps.py will output the current time, latitude, and longitude.  This could be changed to just do it once.

Here is gdpssetup.sh:
```
#!/bin/sh
# steps from https://learn.adafruit.com/adafruit-ultimate-gps-on-the-raspberry-pi/setting-everything-up
# except that the GPS is on serial0 not the ttyUSB0 port
# https://learn.adafruit.com/adafruit-ultimate-gps-hat-for-raspberry-pi/basic-test is another good resource

echo GPS Hat status LED Blinking on and off once every other second - NO FIX
echo GPS Hat status LED Blinking once per 10 seconds or so - HAS FIX

echo You may have to re-run this script if the GPS is not locked.

# these next steps clean up if you have had to run the script before
sudo killall gpsd

sudo systemctl stop gpsd.socket
sudo systemctl disable gpsd.socket

# now run the daemon that parses the GPS NMEA 0183 serial data
sudo gpsd /dev/serial0 -F /var/run/gpsd.sock

echo Test with
echo "    cgps -s"
echo and/or
echo "    gpsmon"
echo It may take a few seconds to see the display update.
```


And here is justgps.py:
```
from gpsdclient import GPSDClient
import board
import busio
from time import sleep

# Listen on port 2947 (gpsd) of localhost
# session = gps.gps("localhost", "2947")
# session.stream(gps.WATCH_ENABLE | gps.WATCH_NEWSTYLE)

client = GPSDClient(host="127.0.0.1")

for result in client.dict_stream(convert_datetime=True):
    if result["class"] == "TPV":
        print("Time: %s" % result.get("time", "n/a"))
        print("Latitude: %s" % result.get("lat", "n/a"))
        print("Longitude: %s" % result.get("lon", "n/a"))
```
